cleverhans/attacks_tf.py

```diff
@@ -417,3 +417,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
 
     # Return augmented training data (needs to be labeled afterwards)
     return X_sub
+
+
+class CarliniWagnerL2:
+    def __init__(self, sess, model, batch_size, confidence,
+                 targeted, learning_rate,
+                 binary_search_steps, max_iterations,
+                 abort_early, initial_const,
+                 clip_min, clip_max, num_labels, shape):
+        """
+        This attack was originally proposed by Carlini and Wagner. It is an
+        iterative attack that finds adversarial examples on many defenses that
+        are robust to other attacks.
+        Paper link: https://arxiv.org/abs/1608.04644
+
+        :param confidence: Confidence of adversarial examples: higher produces
+                           examples that are farther away, but more strongly
+                           classified as adversarial.
+        :param batch_size: Number of attacks to run simultaneously.
+        :param targeted: True if we should perform a targetted attack, False
+                         otherwise.
+        :param learning_rate: The learning rate for the attack algorithm.
```

Comment:
See above, and this probably applies to most of the docstring since it is shared w/ the wrapper class in `attacks.py`